iOS/Flows/Conversation/Threads/BottomToTopColumnCollectionViewLayout.swift

```diff
@@ -0,0 +1,133 @@
+
+                let attributes = UICollectionViewLayoutAttributes(forCellWith: indexPath)
+                attributes.frame = self.frameForItem(at: indexPath)
+                self.cellLayoutAttributes[indexPath] = attributes
+            }
+        }
+    }
+
+    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        // Return all items whose frames intersect with the given rect.
+        return self.cellLayoutAttributes.values.filter { attributes in
+            rect.intersects(attributes.frame)
+        }
+    }
+
+    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
+        // If the attributes are cached already, just return those.
+        if self.cellLayoutAttributes[indexPath] != nil {
+            return self.cellLayoutAttributes[indexPath]
+        }
+
+        let frame = self.frameForItem(at: indexPath)
+        let attributes = UICollectionViewLayoutAttributes(forCellWith: indexPath)
+        attributes.frame = frame
+
+        return attributes
+    }
+
+    // MARK: - Layout calculations
+
+    /// Get the layout attributes for the first item in the collection. This should be the bottom most item.
+    private func firstItemLayoutAttributes() -> UICollectionViewLayoutAttributes? {
+        guard let collectionView = self.collectionView else { return nil }
+
+        // If we don't have a first item, return nil.
+        guard collectionView.numberOfSections > 0, collectionView.numberOfItems(inSection: 0) > 0 else {
+            return nil
+        }
+
+        return self.layoutAttributesForItem(at: IndexPath(item: 0, section: 0))
+    }
+
+    /// Gets the frame rect for the item at the given index.
+    private func frameForItem(at indexPath: IndexPath) -> CGRect {
+        let origin = self.originForItem(at: indexPath)
+        let size = CGSize(width: self.itemSize.width, height: self.itemSize.height)
+
+        return CGRect(origin: origin, size: size)
+    }
+
+    /// Gets the top left corner position for the item at the given index path.
+    private func originForItem(at index: IndexPath) -> CGPoint {
+        guard let collectionView = self.collectionView else { return .zero }
+
+        // Get the height of all the sections above this one.
+        var heightOfPreviousSections: CGFloat = 0
+        for sectionIndex in index.section..<collectionView.numberOfSections {
+            heightOfPreviousSections += self.heightOfSection(sectionIndex)
+        }
+
+        // Get the number items above this item to determine how down the item should be pushed.
+        // The lowest index will have the most items on top, and the highest index will be on the top.
+        let itemsAboveCount = CGFloat(collectionView.numberOfItems(inSection: 0) - index.item - 1)
+        let y = itemsAboveCount * (self.itemSize.height + self.itemSpacing)
+        return CGPoint(x: 0, y: y)
+    }
```

Comment:
This is the where most of the layout logic happens. Long story short, we put the first cell at the bottom of the column.